refactor(rust/treeshaking): get resolved symbol of `MemberExprRef` in one go

hyf0 opened the pull request:
<!-- Thank you for contributing! -->

### Description

- This PR does does the same thing as the old code but storing the result in a map and use it directly rather do a resolving process with cache.
- This PR also split the member expression resolving logic into a extra step, it might slow down the performance but make the treeshaking process more clean and simple.
- https://github.com/rolldown/rolldown/pull/1805 rely on this PR, which makes the finalizing logic simpler.

<!-- Please insert your description here and provide especially info about the "what" this PR is solving -->

hyf0 commented:
* **#1818** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/1818?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#1816** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/1816?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1815** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/1815?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1814** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/1814?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1813** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/1813?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @hyf0 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

hyf0 commented on crates/rolldown_common/src/types/member_expr_ref.rs:
```diff
@@ -9,4 +10,32 @@ use crate::SymbolRef;
+  pub fn resolved_symbol_ref(
+    &self,
+    resolved_map: &FxHashMap<Span, Option<(SymbolRef, Vec<CompactStr>)>>,
+  ) -> Option<SymbolRef> {
```
return `None` need to be handled specially in code splitting and treeshaking. The old code fallback to `object_ref`, which is actually negetive.

IWANABETHATGUY commented on crates/rolldown/src/ast_scanner/impl_visit.rs:
```diff
@@ -58,14 +59,17 @@ impl<'me, 'ast> Visit<'ast> for AstScanner<'me> {
+          // - Import statements are hoisted to the top of the module, so in this time being, all imports are scanned.
+          // - Having empty span will also results to bailout since we rely on span to identify ast nodes.
+          Some(sym_ref)
+            if self.result.named_imports.contains_key(&sym_ref) && !expr.span().is_unspanned() =>
```
What if other plugin inserted a top-level `MemberExpr` in **transform** phase?

hyf0 replied:
> Import statements are hoisted to the top of the module, so in this time being, all imports are scanned.

No matter what transform, this rule still applies.

> except that using span as a key might be a footgun in the future.

As for invalid span, we just do bailout. We'll switch `AstNoeId` once oxc stablize it.

IWANABETHATGUY replied:
`AstNodeId` sounds more robust

IWANABETHATGUY replied:
I saw a litlte issue when I try to inserted a ZERO SPAN member expr, e.g. 
```js
// MULTIPLE ENTRY MODULES
import * as test from './hyper-cube.js';

test.a()
test.b() // inserted by transformer.
// hyper-cube.js
export const a = 100;
export const b = 1000;

```
It would ref the whole namespace.
**output** 
```js

//#region rolldown:runtime
var __defProp = Object.defineProperty;
var __export = (target, all) => {
	for (var name in all) __defProp(target, name, {
		get: all[name],
		enumerable: true
	});
};

//#endregion
//#region hyper-cube.js
var hyper_cube_ns = {};
__export(hyper_cube_ns, {
	a: () => a,
	b: () => b
});
const a = 100;
const b = 1000;

//#endregion
//#region entry.js
a();
hyper_cube_ns.b();

//#endregion
//# sourceMappingURL=entry.js.map
```


**expected** 
```js

//#region hyper-cube.js
const a = 100;
const b = 1000;

//#endregion
//#region entry.js
a();
b();

//#endregion
//# sourceMappingURL=entry.js.map
```

IWANABETHATGUY replied:
the result is still right, just wait for `AstNodeId` stablized.

hyf0 replied:
Yeah. It's expected. We have code to prevent scanning such member expr by `!expr.span().is_unspanned()`.

IWANABETHATGUY commented:
Overall, LGTM, except that using span as a key might be a footgun in the future.

hyf0 committed:
refactor(rust/treeshaking): get resolved symbol of `MemberExprRef` in one go (#1818)

<!-- Thank you for contributing! -->

### Description

- This PR does does the same thing as the old code but storing the result in a map and use it directly rather do a resolving process with cache.
- This PR also split the member expression resolving logic into a extra step, it might slow down the performance but make the treeshaking process more clean and simple.
- https://github.com/rolldown/rolldown/pull/1805 rely on this PR, which makes the finalizing logic simpler.

<!-- Please insert your description here and provide especially info about the "what" this PR is solving -->
